include/exadg/convection_diffusion/driver.cpp

```diff
@@ -143,28 +144,124 @@ Driver<dim, Number>::ale_update() const
   time_int_bdf->ale_update();
 }
 
+template<int dim, typename Number>
+void
+Driver<dim, Number>::mark_cells_coarsening_and_refinement(dealii::Triangulation<dim> & tria,
+                                                          VectorType const & solution) const
+{
+  mark_cells_kelly_error_estimator(tria,
+                                   pde_operator->get_dof_handler(),
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  // Update mapping
+  AssertThrow(ale_mapping.get() == 0,
+              dealii::ExcMessage(
+                "Combination of adaptive mesh refinement and ALE not implemented."));
+
+  std::shared_ptr<dealii::MappingQCache<dim>> mapping_q_cache =
+    std::dynamic_pointer_cast<dealii::MappingQCache<dim>>(mapping);
+  AssertThrow(
+    mapping_q_cache.get() == 0,
+    dealii::ExcMessage(
+      "Combination of adaptive mesh refinement and dealii::MappingQCache not implemented."));
+
+  pde_operator->setup_after_coarsening_and_refinement();
+
+  postprocessor->setup(*pde_operator);
+}
+
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::do_adaptive_refinement(unsigned int const time_step_number)
+{
+  // AMR is only implemented for implicit timestepping.
+  if(trigger_coarsening_and_refinement_now(
+       application->get_parameters().amr_data.trigger_every_n_time_steps, time_step_number))
+  {
+    std::shared_ptr<TimeIntBDF<dim, Number>> bdf_time_integrator =
+      std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+    AssertThrow(bdf_time_integrator.get(),
+                dealii::ExcMessage("Combination of adaptive mesh refinement "
+                                   "and explicit time integration not implemented."));
```

Comment:
Is this check needed at all? Below, we have the call `time_integrator->prepare_coarsening_and_refinement();`. Isn't this the place where one would check this?

Reply:
The point is that neither the explicit time integrator nor the corresponding base class have the `get_solution_np`, which gives a solution vector at $t^{n+1}$ to base the adaptive refinement on. That is the reason for the cast.
I did not introduce `get_solution_np()` in the base class, but I could? Once we have `time_integrator_rk->get_solution_np()`, I can remove the cast here completely. WIth this, the check for a successful cast would also be removed! ... I would rather not add explicit timestepping in this PR, ok?